networkx/algorithms/tests/test_lowest_common_ancestors.py

```diff
@@ -133,6 +136,7 @@ def test_tree_all_pairs_lowest_common_ancestor12(self):
         G = nx.DiGraph([(3, 4), (5, 4)])
         pytest.raises(nx.NetworkXError, list, tree_all_pairs_lca(G))
 
+    # HOW TO PARAMETRIZE THIS ONE?
```

Comment:
@rossbar do you have any magic to handle parametrizing this test method?

Reply:
I would say that for this test it makes more sense to parametrize on the *inputs* since some of the tested objects are generators (i.e. some require `next` in order to trigger there behavior) and some are functions. For example, I'd probably write this test like so:

```python
@pytest.mark.parametrize("graph_type", (nx.Graph, nx.MultiGraph, nx.MultiDiGraph))
def test_not_implemented(self, graph_type):
    G = graph_type([(0, 1)])
    with pytest.raises(nx.NetworkxNotImplemented):
        next(tree_all_pairs_lca(G))
    with pytest.raises(nx.NetworkxNotImplemented):
        next(all_pairs_lca(G))
    with pytest.raises(nx.NetworkxNotImplemented):
        nx.lowest_common_ancestor(G, 0, 1)
```

... and if you wanted to add more generators/functions, you could add more `with` statements (e.g.

```python
with pytest.raises(nx.NetworkxNotImplemented):
    nx.naive_lowest_common_ancestor(G, 0, 1)
```

The other thing you could do is split this up into two separate tests, one for generators and one for functions, that would allow you to [stacked parametrizations (see the end of this section)](https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions). That might look something like:

```python
@pytest.mark.parametrize("graph_type", (nx.Graph, nx.MultiGraph, nx.MultiDiGraph))
@pytest.mark.parametrize("lca_gen", (tree_all_pairs_lca, all_pairs_lca))
def test_not_implemented_generator(self, graph_type, lca_gen):
    G = graph_type([(0, 1)])
    with pytest.raises(nx.NetworkxNotImplemented):
        next(lca_gen(G))

@pytest.mark.parametrize("graph_type", (nx.Graph, nx.MultiGraph, nx.MultiDiGraph))
@pytest.mark.parametrize("lca_fn", (tree_all_pairs_lca, all_pairs_lca))
def test_not_implemented_function(self, graph_type, lca_fn):
    G = graph_type([(0, 1)])
    with pytest.raises(nx.NetworkxNotImplemented):
        lca_fn(G, 0, 1)
```

The trick with parametrizing tests (IMO) is to strike a balance between utility and readability - of course, where that balance lies is usually pretty subjective :)